core/src/main/java/jenkins/security/SuppressionFilter.java

```diff
@@ -0,0 +1,160 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.security;
+
+import hudson.init.InitMilestone;
+import hudson.init.Initializer;
+import hudson.util.PluginServletFilter;
+import jenkins.model.Jenkins;
+import org.acegisecurity.AccessDeniedException;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.HttpResponses;
+import org.kohsuke.stapler.Stapler;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.nio.charset.StandardCharsets;
+import java.util.UUID;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
+
+@Restricted(NoExternalUse.class)
+public class SuppressionFilter implements Filter {
+
+    private static final Logger LOGGER = Logger.getLogger(SuppressionFilter.class.getName());
+
+    @Initializer(after = InitMilestone.STARTED)
+    public static void init() throws ServletException {
+        HttpResponses.setErrorDetailsFilter((code, cause) -> showStackTrace());
+        PluginServletFilter.addFilter(new SuppressionFilter());
+    }
+
+    public void init(FilterConfig filterConfig) {
+        // no-op
+    }
+
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        try {
+            chain.doFilter(request, response);
+        } catch (Exception e) {
+            if (containsAccessDeniedException(e)) {
+                throw e;
+            }
+            if (Stapler.isSocketException(e)) {
+                return;
+            }
+
+            if (showStackTrace()) {
+                if (e instanceof IOException || e instanceof ServletException || e instanceof RuntimeException) {
+                    // thing we can throw without wrapping
+                    throw e;
+                }
+                throwServletException(e);
+            } else if (request instanceof HttpServletRequest && response instanceof HttpServletResponse) {
+                String errorId = logException((HttpServletRequest) request, e);
```

Comment:
will cause a lot of log spam for things like client disconnections during write (which are to be considered normal).

Reply:
Possibly. I can't recall seeing any stack traces in situations like that in the past. It could be that they don't show because of the type of disconnection. I'm not sure how to reproduce these types of problems in a meaningful way.

We could not log the exceptions. Just swallow them and present the error page. Or log them in at a different level. Or if we can detect these types of failures we could handle them differently.

If we could get some idea of what exceptions might fall into this category and how to examine them, we can see if there's something special we need to do here.

Reply:
Maybe `Stapler.isSocketException(e)`, currently on line 87, covers that case.